homeassistant/components/switch/ads.py

```diff
@@ -0,0 +1,76 @@
+"""Support for ADS switch platform."""
+
+import logging
+
+import voluptuous as vol
+from homeassistant.components.switch import PLATFORM_SCHEMA
+from homeassistant.const import CONF_NAME
+from homeassistant.components.ads import DATA_ADS
+from homeassistant.helpers.entity import ToggleEntity
+import homeassistant.helpers.config_validation as cv
+
+
+_LOGGER = logging.getLogger(__name__)
+DEPENDENCIES = ['ads']
+DEFAULT_NAME = 'ADS Switch'
+CONF_ADSVAR = 'adsvar'
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_ADSVAR): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up switch platform for ADS."""
+    ads_hub = hass.data.get(DATA_ADS)
+    if not ads_hub:
+        return False
+
+    dev_name = config.get(CONF_NAME)
+    ads_var = config.get(CONF_ADSVAR)
+
+    add_devices([AdsSwitch(ads_hub, dev_name, ads_var)], True)
+
+
+class AdsSwitch(ToggleEntity):
+    """Representation of an Ads switch device."""
+
+    def __init__(self, ads_hub, dev_name, ads_var):
+        """Initialize the AdsSwitch entity."""
+        self._ads_hub = ads_hub
+        self._on_state = False
+        self.dev_name = dev_name
+        self.ads_var = ads_var
+
+    @property
+    def is_on(self):
+        """Return if the switch is turned on."""
+        return self._on_state
+
+    @property
+    def name(self):
+        """Return the name of the entity."""
+        return self.dev_name
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self._ads_hub.write_by_name(self.ads_var, True,
+                                    self._ads_hub.PLCTYPE_BOOL)
+        self._on_state = True
+
+    def turn_off(self, **kwargs):
+        """Turn the switch off."""
+        self._ads_hub.write_by_name(self.ads_var, False,
+                                    self._ads_hub.PLCTYPE_BOOL)
+        self._on_state = False
```

Comment:
Remove this. If you have a pooling device, we call the update automatic and you have a callback that update it automatic?

Reply:
Thanks for your review. Could you please explain shortly why you want me to remove the `turn_off()` function? It is needed to pass the changed switch state to the component which then communicates with the actual device. I can not see what is wrong about this, respectively how to handle it differently. I'm surely missing something here and maybe you can give me a hint what you're up to.

My second question concerns the communication from component to entity. If I put it right you want me to replace the callbacks by some helper function from `helpers.dispatcher`. I couldn't find any proper documentation on `helpers.dispatcher` in the API. I will have a closer look at the source code on Github but maybe you could provide me a link or a code sample that helps me implement it the right way.

Regards

Reply:
It give 2 kind of device. First is polling, that mean we ask for device status ever x seconds and as the new state after we call the service. Second is with a callback, that mean the device push all device updates to entity. A component should be select one of that to recieve data.

You need remove `self._on_state` a review is only for the last line...

Reply:
I'm not sure if your refer to the communication with the **device** itself or if you mean communication between **component** and **entity**. 

Actually the component is designed to allow polling as well as the use of device notifications (via callback) to get information from the **device**. I have added a config parameter `use_notify` where the user can change between the two modes. I don't see any harm in this in fact it is very useful to adapt the communication with the physical device according to the usecase.

However I understand you might refer to the communication between **component** and **entity** respectively the communication within homeassistant where I mix things up. So I will have a look at the communication there and come back to you. Maybe just give me a short hint if I am on the right track.

Reply:
Don't let user deside what he want. Implement push or pull but not both, that make the code to complex. If you use push (device to home-assistant) lets implement the callback on platform so you don't need communication from component to platform.

Reply:
@pvizeli I removed the Polling from all platforms. I hope this fits now. 

I decided to preprocess the callbacks by the hub component in order to keep device-specific code away from the platform code. I wouldn't want to put the typecasts in the platform code. 

So I figured out I had two choices:

- handle the device notifications in the Hub component, do the casting and forward them to the platform entities or
- create a new EntityClass that contains the typecasts and that I derive from. 

I decided for the first way because it keeps all device notifications in the same place. Also I need to send a signal to my device in case I shut down Homeassistant that tells my device to disable the notification.

Reply:
@pvizeli I made plenty of changes since your last review and tried to implement your suggestions. Could you please have another look at the PR and approve if it's alright now.